text/0036-async-testbench-functions.md

```diff
@@ -0,0 +1,236 @@
+    await test_rgb(sim, 0, 0, 0)
+    await test_rgb(sim, 255, 0, 0)
+    await test_rgb(sim, 0, 255, 0)
+    await test_rgb(sim, 0, 0, 255)
+    await test_rgb(sim, 255, 255, 255)
+```
+
+Since `.send()` and `.recv()` invokes `sim.get()` and `sim.set()` that in turn will invoke the appropriate value conversions for a value castable (here `data.View`), it is general enough to work for streams with arbitrary shapes.
+
+`Tick()` and `Delay()` are replaced by `sim.tick()` and `sim.delay()` respectively.
+In addition, `sim.changed()` and `sim.edge()` is introduced that allows creating triggers from arbitrary signals.
+These all return a trigger object that can be made conditional through `.until()`.
+
+`Active()` and `Passive()` are replaced by an `background=False` keyword argument to `.add_testbench()`.
+Processes created through `.add_process()` are always created as background processes.
+To allow a background process to ensure an operation is finished before end of simulation, `sim.critical()` is introduced, which is used as a context manager:
+
+```python
+async def packet_reader(sim, stream):
+    while True:
+        # Wait until stream has valid data.
+        await sim.tick().until(stream.valid)
+
+        # Ensure simulation doesn't end in the middle of a packet.
+        async with sim.critical():
+            packet = await stream.read_packet()
+            print('Received packet:', packet.hex(' '))
+```
+
+When a trigger object is awaited, it'll return the value(s) of the trigger(s), and it can also be used as an async generator to repeatedly await the same trigger.
+Multiple triggers can be combined.
+Consider the following examples:
+
+Combinational adder as a process:
+```python
+a = Signal(); b = Signal(); o = Signal()
+async def adder(sim):
+    async for a_val, b_val in sim.changed(a, b):
+        await sim.set(o, a_val + b_val)
+sim.add_process(adder)
+```
+
+DDR IO buffer as a process:
+```python
+o = Signal(2); pin = Signal()
+async def ddr_buffer(sim):
+    while True: # could be extended to pre-capture next `o` on posedge
+        await sim.negedge()
+        await sim.set(pin, o[0])
+        await sim.posedge()
+        await sim.set(pin, o[1])
+sim.add_process(ddr_buffer)
+```
+
+Flop with configurable edge reset and posedge clock as a process:
+```python
+clk = Signal(); rst = Signal(); d = Signal(); q = Signal()
+def dff(rst_edge):
+    async def process(sim):
+        async for clk_hit, rst_hit in sim.posedge(clk).edge(rst, rst_edge):
+            await sim.set(q, 0 if rst_hit else await sim.get(d))
+    return process
+sim.add_process(dff(rst_edge=0))
+```
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following `Simulator` methods have their signatures updated:
+
+* `add_process(process)`
+* `add_testbench(process, *, background=False)`
+
+The new optional named argument `background` registers the testbench as a background process when true.
```

Comment:
This doesn't explain when processes created by `add_process()` become background or not. They become background iff an async generator function is passed as `process`, which is quite not obvious!